### PR TITLE
enhance: adds truststores to the keycloak cr

### LINKED ISSUE #24798: Add truststores to keycloak cr
### Description

For #23742 the Keycloak cr will be enhanced to support the truststore-paths option.  Based upon the server support, the cr change should be as simple as:

```
spec:
  ...
  trustStores:
    - name: my-truststore
      secret: my-secret 
    ...
 ```

We could also support configmaps in addition to secrets.

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by shawkins**
@vmuzikar there are a couple of options for the cr design.  The choice here was to do the minimum:

```
truststores:
  foo: {}
  bar: {}
```

- where the empty object typically wouldn't be used, but I did put an entry for optional handling in there for now.  It could eventually handle the other aspects of something like SecretVolumeSource.
- rather than an array with an effective name key (which will current lack the merge key in the crd), I'm just using a map with the assumption that the key is the secret name.  This also means that there's no choice given to the user in terms of defining a mount path, or an ordering of the entries - both of those seem fine given our current assumptions.  Lastly this could eventually allow for configMaps, you'd just have to use a qualified name - configMap/foo: {} - but that handling was not added as we don't currently watch configMaps.

As shown in the issue, the alternative is: 

```
truststores:
  - name: foo
    secret: some-name
```

This is more inline with what is done in kube currently.  The drawbacks are:
- the name isn't meaningful at this point (we cannot reflect that it's a merge key as the crd generator doesn't have that feature) and we may not want it to be in terms of a mount name or path.  
- if we want to support configmaps, that won't be well reflected in the crd - we'll have secret and configMap as siblings, and will need to validate on our own that only 1 is present.

closes: #24798

### CODE REVIEW

**Review by ghost — COMMENTED** [review]
Unreported flaky test detected, please review

**Comment by vmuzikar** [comment]
> The choice here was to do the minimum:
> 
> ```
> truststores:
>   foo: {}
>   bar: {}
> ```

TBH, this doesn't seem very intuitive to me. I was thinking about it a bit more and IMHO we could try to provide similar UX to what we have for CLI where we have `truststore-paths`. Perhaps we could have something like `truststore.secrets`, i.e.:
```
truststore:
  secrets:
    - mySecret1
    - mySecret2
```
That'd allow us to add support for additional truststore sources (like CMs) in the future if needed.

To be more future proof to allow for additional fields (like optional), we could do:
```
truststore:
  secrets:
    - name: mySecret1
    - name: mySecret2
```

WDYT?

**Comment by vmuzikar** [comment]
Thanks for the update.

I wonder if we need the `truststores.name` field. It's not the key now, so technically not needed? Or we could make it a map key:
```
truststores:
  foo:
    secret:
      name: foo
```
which would solve the missing patch merge key issue.

Or we could omit the name whatsoever:
```
truststores:
  - secret:
      name: foo
```

**Comment by shawkins** [comment]
> I wonder if we need the truststores.name field. It's not the key now, so technically not needed?

It's not, again I'm just doing this for consistency with an existing paradigm.

> Or we could make it a map key:

Sure, I'll just let others weigh in before making any other changes.

> Or we could omit the name whatsoever:

If we ever want to support the merge key, this won't work.  As far as I know you have to reference a direct child property.

**Comment by vmuzikar** [comment]
> As far as I know you have to reference a direct child property.

I see, didn't realize that. Then there's no avoiding of having the arbitrary name anyway (unless we want to go the empty objects way where the secret name is the map key). Either as a map key, or a child field. In that case I'm more inclined towards the map, this will at least avoid the merge key issue.

**Comment by Pepo48** [comment]
>  In that case I'm more inclined towards the map, this will at least avoid the merge key issue.

As per yesterday's discussion, I probably also favour this option after all.

**Comment by vmuzikar** [comment]
> As per yesterday's discussion, I probably also favour this option after all.

@Pepo48 Just to clarify, you mean this one
```
truststores:
  foo: {}
  bar: {}
```
or this one? :)
```
truststores:
  foo:
    secret:
      name: foo
```

**Comment by Pepo48** [comment]
@vmuzikar sorry for not being clear enough - the second one after the arguments you stated yesterday 😊 (even though I liked the first option, I agree that it might not be that intuitive).

**Comment by shawkins** [comment]
Created an upstream issue https://github.com/fabric8io/kubernetes-client/issues/5639 - there are a couple of openapi directives kubernetes looks at to control the patching behavior.

So it sounds like we're leaning toward:

```
truststores:
  foo:
    secret:
      name: foo
```

Instead of:

```
truststores:
  - name: foo
    secret:
      name: foo
```

Correct?

**Comment by vmuzikar** [comment]
> Correct?

Correct.

@mabartos Do you want to take a vote? :)

**Comment by shawkins** [comment]
I'll update the pr to use a top level map and move it out of draft.

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/TruststoreSource.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@Buildable(editableEnabled = false, builderPackage = "io.fabric8.kubernetes.api.builder")
+public class TruststoreSource {
+
+    @Required
+    private String name;
+    private Boolean optional;
+
+    public Boolean getOptional() {
```
Comment (mabartos) [inline operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/TruststoreSource.java:33]:
```suggestion
    public Boolean isOptional() {
```
Better, right?

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/Truststore.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package org.keycloak.operator.crds.v2alpha1.deployment.spec;
```
Comment (mabartos) [inline operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/Truststore.java:1]: Could you please add the license?

File: docs/documentation/release_notes/topics/24_0_0.adoc [hunk 3]
```diff
@@ -6,6 +6,24 @@ The Keycloak JS adapter now uses the https://webpack.js.org/guides/package-expor
...
 Keycloak introduces an improved truststores configuration options. The Keycloak truststore is now used across the server: for outgoing connections, mTLS, database drivers and more. It's no longer needed to configure separate truststores for individual areas. To configure the truststore, you can put your truststores files or certificates in the default `conf/truststores`, or use the new `truststore-paths` config option. For details refer to the relevant https://www.keycloak.org/server/keycloak-truststore[guide].
 
+== Keycloak CR Truststores
+
+You may also take advantage of the new server-side handling of truststores via the Keycloak CR, for example:
+
+[source,yaml]
+----
+spec:
+  trustores:
```
Comment (mabartos) [inline docs/documentation/release_notes/topics/24_0_0.adoc:16]:
```suggestion
  truststores:
```

File: operator/src/test/java/org/keycloak/operator/testsuite/integration/KeycloakTruststoresTests.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.keycloak.operator.testsuite.utils.K8sUtils.deployKeycloak;
+import static org.keycloak.operator.testsuite.utils.K8sUtils.getResourceFromFile;
+
+@QuarkusTest
+public class KeycloakTruststoresTests extends BaseOperatorTest {
+    @Test
+    public void testTrustoreMissing() {
```
Comment (mabartos) [inline operator/src/test/java/org/keycloak/operator/testsuite/integration/KeycloakTruststoresTests.java:39]:
```suggestion
    public void testTruststoreMissing() {
```

File: operator/src/test/java/org/keycloak/operator/testsuite/integration/KeycloakTruststoresTests.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            StatefulSet statefulSet = stsResource.get();
+            assertEquals("true",
+                    statefulSet.getMetadata().getAnnotations().get(Constants.KEYCLOAK_MISSING_SECRETS_ANNOTATION));
+            assertTrue(statefulSet.getMetadata().getAnnotations().get(Constants.KEYCLOAK_WATCHING_ANNOTATION)
+                    .contains("xyz"));
+        });
+    }
+
+    @Test
+    public void testTrustroreExists() {
```
Comment (mabartos) [inline operator/src/test/java/org/keycloak/operator/testsuite/integration/KeycloakTruststoresTests.java:56]:
```suggestion
    public void testTruststoreExists() {
```

**Review by mabartos — CHANGES_REQUESTED** [review]
@shawkins I did a quick look and some things should be assessed.

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/TruststoreSource.java (hunk 1 above)
Comment (shawkins) [reply operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/TruststoreSource.java:33]: No, "is" should only be for the primitive boolean.

**Comment by shawkins** [comment]
@mabartos thanks for catching the misspelling - I've generally corrected that in the codebase as it appears like a common mistake

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/Truststore.java (hunk 2 above)
Comment (shawkins) [reply operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/Truststore.java:18]: Added now. Just to double check on this. There doesn't seem to be consistency (it's missing in quite a few source files), or an automated check to enforce that it's included. I know we require one in fabric8, but it's a dumb check so we're forced to use the same copyright year for everything. It also appears in keycloak that the copyright year is set for each file, is that correct as well?

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/TruststoreSource.java (hunk 1 above)
Comment (mabartos) [reply operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/TruststoreSource.java:33]: @shawkins yep, you're right

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/Truststore.java (hunk 2 above)
Comment (mabartos) [reply operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/Truststore.java:18]: Like, I don't know the license requirements for the Keycloak code, but AFAIK, it's been a general convention to put the license to new classes. But it's not a hard requirement, I guess. It just blicked to my eye :)

**Review by mabartos — APPROVED** [review]
@shawkins LGTM. The test coverage should be sufficient.

**Review by abstractj — APPROVED** [review]

**Pull request merged**